dali/operators/sequence/optical_flow/optical_flow_impl/optical_flow_impl.cc

```diff
@@ -0,0 +1,347 @@
+  width_max_ = GetCapabilities(of_inst_.nvOFGetCaps, of_handle_, NV_OF_CAPS_WIDTH_MAX);
+  height_max_ = GetCapabilities(of_inst_.nvOFGetCaps, of_handle_, NV_OF_CAPS_HEIGHT_MAX);
+  IsSizeSupported(width_min_, width_max_, width_, "Width");
+  IsSizeSupported(height_min_, height_max_, height_, "Height");
+
+  CUDA_CALL(of_inst_.nvOFSetIOCudaStreams(of_handle_, stream_, stream_));
+  auto status = VerifySupport(of_inst_.nvOFInit(of_handle_, &init_params_));
+  CUDA_CALL(status);
+
+  CreateBuffers();
+}
+
+void OpticalFlowImpl::CreateBuffers() {
+  inbuf_.reset(
+          new OpticalFlowBuffer(of_handle_, width_, height_, of_inst_, NV_OF_BUFFER_USAGE_INPUT,
+                                NV_OF_BUFFER_FORMAT_ABGR8));
+  refbuf_.reset(
+          new OpticalFlowBuffer(of_handle_, width_, height_, of_inst_, NV_OF_BUFFER_USAGE_INPUT,
+                                NV_OF_BUFFER_FORMAT_ABGR8));
+  outbuf_.reset(
+          new OpticalFlowBuffer(of_handle_, div_ceil(width_, out_grid_size_),
+                                div_ceil(height_, out_grid_size_), of_inst_,
+                                NV_OF_BUFFER_USAGE_OUTPUT, NV_OF_BUFFER_FORMAT_SHORT2));
+  if (of_params_.enable_external_hints) {
+    hintsbuf_.reset(
+            new OpticalFlowBuffer(of_handle_, div_ceil(width_, hint_grid_size_),
+                                  div_ceil(height_, hint_grid_size_), of_inst_,
+                                  NV_OF_BUFFER_USAGE_HINT, NV_OF_BUFFER_FORMAT_SHORT2));
+  }
+}
+
+void OpticalFlowImpl::DestroyBuffers() {
+  inbuf_.reset(nullptr);
+  refbuf_.reset(nullptr);
+  outbuf_.reset(nullptr);
+  if (of_params_.enable_external_hints) {
+    hintsbuf_.reset(nullptr);
+  }
+}
+
+void OpticalFlowImpl::CreateOf() {
+  DeviceGuard g(device_id_);
+  auto ret = of_inst_.nvCreateOpticalFlowCuda(context_, &of_handle_);
+  if (ret != NV_OF_SUCCESS) {
+    throw unsupported_exception(
+        "Failed to create Optical Flow context: Verify that your device supports Optical Flow.");
+  }
+}
+
+void OpticalFlowImpl::DestroyOf() {
+  if (of_handle_) {
+    auto err = of_inst_.nvOFDestroy(of_handle_);
+    // unload lib no matter if it was successful
+    if (err != NV_OF_SUCCESS) {
+      // Failing to destroy OF leads to significant GPU resource leak,
+      // so we'd rather terminate than live with this
+      std::cerr << "Fatal error: failed to destroy optical flow" << std::endl;
+      std::terminate();
+    }
+  }
+  of_handle_ = nullptr;
+}
+
+void OpticalFlowImpl::Prepare(size_t width, size_t height) {
+  if (width == width_ && height == height_)
+    return;
+  if (width != width_) {
+    width_ = width;
+    IsSizeSupported(width_min_, width_max_, width_, "Width");
+  }
+  if (height != height_) {
+    height_ = height;
+    IsSizeSupported(height_min_, height_max_, height_, "Height");
+  }
+
+  init_params_.width = static_cast<uint32_t>(width_);
+  init_params_.height = static_cast<uint32_t>(height_);
+
+  DestroyBuffers();
```

Comment:
Long term we can think about creating multiple of NV OF instances for requested resolutions, but this way we can easily run of out of the memory and it is hard to tell who can benefit from such optimization.